### PR TITLE
PVP-141

### PR DESCRIPTION
**Pull request by vytjar**
Added a step counter to the daily view.

- Step counter is displayed only if the Day composable is supplied by the date
- The step progress is displayed in a circle. The step goal is currently hardcoded as 10000. Later we might need to add a way to set a step goal?
- Added a new method inside HealthConnectService for aggregating steps (the old method was prone to duplicates inside the data, although I left it as it allows to read other types of data besides steps)
- Daily view box size was increased so that the step progress circle fits

This is how the step counter looks in dark mode:

Screenshot_20240314_161726

### CODE REVIEW

File: app/src/main/java/com/pvp/app/ui/screen/calendar/CalendarScreen.kt [hunk 1]
```diff
@@ -209,41 +218,33 @@ fun Day(
...
-                Text(
-                    "Steps",
-                    textAlign = TextAlign.Center,
-                    modifier = Modifier
-                        .fillMaxWidth()
-                        .padding(16.dp)
-                )
-
-                // Don't display date if no value was supplied
+                // Don't display steps if no date was supplied
```
Comment (klaidoshka) [inline app/src/main/java/com/pvp/app/ui/screen/calendar/CalendarScreen.kt:235]: Comment seems to be redundant. I think it is quite obvious from the code.

File: app/src/main/java/com/pvp/app/ui/screen/calendar/CalendarScreen.kt [hunk 2]
```diff
@@ -291,6 +292,80 @@ enum class TaskFilter(val displayName: String) {
...
+        } else {
+            launcher.launch(PERMISSIONS)
+        }
+    }
+
+    val steps = model.stepsCount.collectAsStateWithLifecycle()
+    val goal = 10000f // TODO create way for user to set a step goal?
+    val progress = steps.value / goal
+
+
```
Comment (klaidoshka) [inline app/src/main/java/com/pvp/app/ui/screen/calendar/CalendarScreen.kt:321]: Redundant single line.

File: app/src/main/java/com/pvp/app/ui/screen/calendar/CalendarScreen.kt [hunk 3]
```diff
@@ -291,6 +292,80 @@ enum class TaskFilter(val displayName: String) {
...
+    Box(
+        contentAlignment = Alignment.Center,
+        modifier = Modifier.fillMaxSize()
+    ) {
+        // These have to be set here because for some reason MaterialTheme.colorScheme
+        // can't be used inside a Canvas ¯\_(ツ)_/¯
```
Comment (klaidoshka) [inline app/src/main/java/com/pvp/app/ui/screen/calendar/CalendarScreen.kt:327]: Comment seems redundant.

If I recall correctly, `colorScheme` is a composable. Compiler should show that exact problem - that a composable must be called within other composable function, which Canvas is not, if I am not mistaken.

Either way, an error should be shown by the compiler, if developer would not understand why this is moved to outer space and would move it in. So, I believe, this is unnecessary.

File: app/src/main/java/com/pvp/app/ui/screen/calendar/CalendarScreen.kt [hunk 4]
```diff
@@ -291,6 +292,80 @@ enum class TaskFilter(val displayName: String) {
...
+                color = progressArcColor,
+                startAngle = -90f,
+                sweepAngle = 360f * progress,
+                useCenter = false,
+                topLeft = topLeft,
+                size = size,
+                style = Stroke(width = strokeWidth, cap = StrokeCap.Round)
+            )
+        }
+        Text(
```
Comment (klaidoshka) [inline app/src/main/java/com/pvp/app/ui/screen/calendar/CalendarScreen.kt:360]: Single empty line required.

File: app/src/main/java/com/pvp/app/ui/screen/calendar/CalendarViewModel.kt [hunk 5]
```diff
@@ -16,20 +21,28 @@ import kotlinx.coroutines.flow.flowOf
...
 
 @HiltViewModel
 @OptIn(ExperimentalCoroutinesApi::class)
 class CalendarViewModel @Inject constructor(
     private val taskService: TaskService,
-    private val userService: UserService
+    private val userService: UserService,
+    private val healtConnectService: HealthConnectService,
+    private val client: HealthConnectClient
+
```
Comment (klaidoshka) [inline app/src/main/java/com/pvp/app/ui/screen/calendar/CalendarViewModel.kt:37]: Redundant empty line.

File: app/src/main/java/com/pvp/app/ui/screen/calendar/CalendarViewModel.kt [hunk 6]
```diff
@@ -60,10 +73,30 @@ class CalendarViewModel @Inject constructor(
...
+    suspend fun permissionsGranted(): Boolean {
+        val granted = client.permissionController.getGrantedPermissions()
+
+        return granted.containsAll(PERMISSIONS)
+    }
+
+    fun getDaysSteps(date: LocalDate) {
+        viewModelScope.launch {
+            val end = date.plusDays(1).atStartOfDay(ZoneId.systemDefault()).toInstant()
+            val start = date.atStartOfDay(ZoneId.systemDefault()).toInstant()
```
Comment (klaidoshka) [inline app/src/main/java/com/pvp/app/ui/screen/calendar/CalendarViewModel.kt:86]: Chained methods should be split.

File: app/src/main/java/com/pvp/app/ui/screen/calendar/CalendarViewModel.kt [hunk 7]
```diff
@@ -60,10 +73,30 @@ class CalendarViewModel @Inject constructor(
...
+    }
 }
 
 data class CalendarState(
     val tasksMonth: List<Task> = listOf(),
     val tasksWeek: List<Task> = listOf(),
     val user: User? = null
+)
+
+val PERMISSIONS = setOf(
```
Comment (klaidoshka) [inline app/src/main/java/com/pvp/app/ui/screen/calendar/CalendarViewModel.kt:99]: I believe this should be lowercased, lint code suggestions by Android Studio should show it too. If they don't show it, means I am wrong and it should stay as it is.

File: app/src/main/java/com/pvp/app/ui/screen/calendar/CalendarScreen.kt [hunk 8]
```diff
@@ -291,6 +292,80 @@ enum class TaskFilter(val displayName: String) {
...
+            val size = Size(radius, radius)
+
+            drawArc(
+                color = backgroundArcColor,
+                startAngle = -90f,
+                sweepAngle = 360f,
+                useCenter = false,
+                topLeft = topLeft,
+                size = size,
+                style = Stroke(width = strokeWidth, cap = StrokeCap.Round)
```
Comment (klaidoshka) [inline app/src/main/java/com/pvp/app/ui/screen/calendar/CalendarScreen.kt:347]: Multiple parameters could be split: here, below and in HealthConnectorServiceImpl. Be consistent.

**Review by klaidoshka — APPROVED** [review]

File: app/src/main/java/com/pvp/app/ui/screen/calendar/CalendarScreen.kt (hunk 3 above)
Comment (vytjar) [reply app/src/main/java/com/pvp/app/ui/screen/calendar/CalendarScreen.kt:327]: Yes, it throws that error, but

image

Doesn't the @Composable mean it's a composable?

Comment (klaidoshka) [reply app/src/main/java/com/pvp/app/ui/screen/calendar/CalendarScreen.kt:327]: Yes, but you are providing code for lambda context, which is not a @Composable, but a simple DrawScope instance method.

If it was `DrawScope.@Composable() -> Unit`, then it should work.

Comment (vytjar) [reply app/src/main/java/com/pvp/app/ui/screen/calendar/CalendarScreen.kt:327]: Ah, makes sense. Thanks

**Pull request merged**